File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/OperationStatus.scala
package org.infinispan.server.hotrod

import scala.annotation.switch

/**
 * Hot Rod operation possible status outcomes.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
object OperationStatus extends Enumeration {
   type OperationStatus = Value

   val Success = Value(0x00)
   val OperationNotExecuted = Value(0x01)
   val KeyDoesNotExist = Value(0x02)
   val SuccessWithPrevious = Value(0x03)
   val NotExecutedWithPrevious = Value(0x04)
   val InvalidIteration = Value(0x05)

   val SuccessCompat = Value(0x06)
   val SuccessWithPreviousCompat = Value(0x07)
   val NotExecutedWithPreviousCompat = Value(0x08)

   val InvalidMagicOrMsgId = Value(0x81)
   val UnknownOperation = Value(0x82)
   val UnknownVersion = Value(0x83) // todo: test
   val ParseError = Value(0x84) // todo: test
   val ServerError = Value(0x85) // todo: test
   val OperationTimedOut = Value(0x86) // todo: test
   val NodeSuspected = Value(0x87)
   val IllegalLifecycleState = Value(0x88)

   def withCompatibility(st: OperationStatus, isCompatibilityEnabled: Boolean): OperationStatus = {
      if (isCompatibilityEnabled) {
         (st: @switch) match {
            case Success => SuccessCompat;
            case SuccessWithPrevious => SuccessWithPreviousCompat;
            case NotExecutedWithPrevious => NotExecutedWithPreviousCompat;
            case _ => st
         }
      } else st
   }
}

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/OperationStatus.java
package org.infinispan.server.hotrod;

import java.util.HashMap;
import java.util.Map;

/**
 * Hot Rod operation possible status outcomes.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
public enum OperationStatus {
   // TODO: need to get codes
   Success(0x00),
   OperationNotExecuted(0x01),
   KeyDoesNotExist(0x02),
   SuccessWithPrevious(0x03),
   NotExecutedWithPrevious(0x04),
   InvalidIteration(0x05),

   SuccessObjStorage(0x06),
   SuccessWithPreviousObjStorage(0x07),
   NotExecutedWithPreviousObjStorage(0x08),

   InvalidMagicOrMsgId(0x81),
   UnknownOperation(0x82),
   UnknownVersion(0x83), // todo: test
   ParseError(0x84), // todo: test
   ServerError(0x85), // todo: test
   OperationTimedOut(0x86), // todo: test
   NodeSuspected(0x87),
   IllegalLifecycleState(0x88),;

   private static final Map<Byte, OperationStatus> intMap = new HashMap<>();

   static {
      for (OperationStatus status : OperationStatus.values()) {
         intMap.put(status.code, status);
      }
   }

   private final byte code;

   OperationStatus(int code) {
      this.code = (byte) code;
   }

   public byte getCode() {
      return code;
   }

   public static OperationStatus fromCode(byte code) {
      return intMap.get(code);
   }

   static OperationStatus withLegacyStorageHint(OperationStatus st, boolean isObjectStorage) {
      if (isObjectStorage) {
         switch (st) {
            case Success:
               return SuccessObjStorage;
            case SuccessWithPrevious:
               return SuccessWithPreviousObjStorage;
            case NotExecutedWithPrevious:
               return NotExecutedWithPreviousObjStorage;
            default:
               return st;
         }
      } else return st;
   }
}
